common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1908 @@
+		return nil, fmt.Errorf("find_tx_attempts_requiring_resend: %w", ErrInvalidChainID)
+	}
+
+	ms.addressStatesLock.Lock()
+	defer ms.addressStatesLock.Unlock()
+	as, ok := ms.addressStates[address]
+	if !ok {
+		return nil, fmt.Errorf("find_tx_attempts_requiring_resend: %w", ErrAddressNotFound)
+	}
+
+	filter := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.TxAttempts == nil || len(tx.TxAttempts) == 0 {
+			return false
+		}
+		attempt := tx.TxAttempts[0]
+		if attempt.State == txmgrtypes.TxAttemptInProgress {
+			return false
+		}
+		if tx.BroadcastAt.After(olderThan) {
+			return false
+		}
+
+		return false
+	}
+	states := []txmgrtypes.TxState{TxUnconfirmed, TxConfirmedMissingReceipt}
+	attempts := as.FetchTxAttempts(states, filter)
+	// sort by sequence ASC, gas_price DESC, gas_tip_cap DESC
+	sort.Slice(attempts, func(i, j int) bool {
+		return (*attempts[i].Tx.Sequence).Int64() < (*attempts[j].Tx.Sequence).Int64()
+	})
+	// LIMIT by maxInFlightTransactions
+	if len(attempts) > int(maxInFlightTransactions) {
+		attempts = attempts[:maxInFlightTransactions]
+	}
+
+	// deep copy the attempts
+	var eAttempts []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	for _, attempt := range attempts {
+		eAttempts = append(eAttempts, ms.deepCopyTxAttempt(attempt.Tx, attempt))
+	}
+
+	return eAttempts, nil
+}
+
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxWithSequence(_ context.Context, fromAddress ADDR, seq SEQ) (*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], error) {
+	ms.addressStatesLock.Lock()
+	defer ms.addressStatesLock.Unlock()
+	as, ok := ms.addressStates[fromAddress]
+	if !ok {
+		return nil, fmt.Errorf("find_tx_with_sequence: %w", ErrAddressNotFound)
+	}
+
+	filter := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.Sequence == nil {
+			return false
+		}
+
+		return (*tx.Sequence).String() == seq.String()
+	}
+	states := []txmgrtypes.TxState{TxConfirmed, TxConfirmedMissingReceipt, TxUnconfirmed}
+	txs := as.FetchTxs(states, filter)
+	if len(txs) == 0 {
+		return nil, nil
+	}
+
+	return ms.deepCopyTx(txs[0]), nil
+}
+
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTransactionsConfirmedInBlockRange(_ context.Context, highBlockNumber, lowBlockNumber int64, chainID CHAIN_ID) ([]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], error) {
+	if ms.chainID.String() != chainID.String() {
+		return nil, fmt.Errorf("find_transactions_confirmed_in_block_range: %w", ErrInvalidChainID)
+	}
+
+	filter := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		if tx.TxAttempts == nil || len(tx.TxAttempts) == 0 {
+			return false
+		}
+		attempt := tx.TxAttempts[0]
+		if attempt.State != txmgrtypes.TxAttemptBroadcast {
+			return false
+		}
+		if attempt.Receipts == nil || len(attempt.Receipts) == 0 {
```

Comment:
nil check is redundant. `len(attempt.Receipts) == 0` is sufficient to check if it's empty